### ISSUE TITLE
Ideas | Plans | Roadmap

### ISSUE THREAD

**Issue by Coretool** [joshgarde/aurora#1]
I think we have to define a clear plan of what we want to do here. 

My idea is to try to copy the way they built the gui on Chrome(ium)OS to NodeOS. To do so, I suggest, thaz we "study" the code they used in chromeOS and port it to NodeOS. What do you think ?!

**Comment by joshgarde** [comment]
I essentially had the same idea - basically reimplementing the same concepts in Chromium OS as node.js + HTML code where possible.

The checkout instructions for Chromium are over at https://dev.chromium.org/developers/how-tos/get-the-code - Just a warning, it's about a 15-10GB download.

**Comment by Coretool** [comment]
Is this the source of the browser or of the OS ? 10 GB is huge for a browser...

**Comment by joshgarde** [comment]
Ash/Aura are actually part of the Chromium browser - they're basically build options you can enable.

You might be able to get the relevant portions of the source here: (without the 10GB+ download)
https://chromium.googlesource.com/chromium/src/

What you're after is `src/ui/aura` and `src/ash`

**Comment by Coretool** [comment]
Yeah, that's what I've downloaded before ^^

**Comment by joshgarde** [comment]
> 10 GB is huge for a browser...

I had the same thought when I was purchasing a VPS to fulfill the system requirements for the build

**Comment by Coretool** [comment]
^^ Let's see what we can get out of this thingy !

**Comment by Coretool** [comment]
Is it just me or is it impossible to clone only the chromium source from this page ?

**Comment by joshgarde** [comment]
I can clone just fine with `git clone https://chromium.googlesource.com/chromium/src`

**Comment by Coretool** [comment]
Yeah... was an error on my side... It works just fine ^^

**Comment by joshgarde** [comment]
Just pushed an initial commit with some of the extracted files I think we'll need. Looking at the [dependency graph](https://raw.githubusercontent.com/joshgarde/aurora/master/ChromeAshViewsAuradependencies.png), looks like we need to tackle Aura first to get it working with EGL.

**Comment by joshgarde** [comment]
Looking at the documentation, looks like ozone already has an implementation we can use.

https://www.chromium.org/developers/design-documents/ozone

**Comment by joshgarde** [comment]
Also, what license do you want to use? The Chromium license looks compatible with most open source licenses. I was personally leaning towards MIT.

**Comment by joshgarde** [comment]
But it's a lot easier just to license under the same licenses as Chromium and ChromiumOS.

**Comment by Coretool** [comment]
I'd use MIT ... That's basically the most open license you can get.

**Comment by Coretool** [comment]
I've looked into ozone and I think I've found what I wanted: 

Ozone has got 3 different platforms. The only one that we're really interested in is probably gbm. Gbm needs mesa so we will have to wait for lite to finish the mesa port. 

Ozone only renders in 600x800 so we might want to scale this since full screen would be much better !

**Comment by joshgarde** [comment]
I was looking at targeting gbm as well and once we figure out how it does the rendering, it shouldn't be too hard to change the rendering size.

The source code from Google uses their own ["configure" system](https://chromium.googlesource.com/chromium/src/tools/gn/) and the [ninja build system](https://ninja-build.org/). Should we keep their stuff or try to port over to autoconf and automake?

**Comment by Coretool** [comment]
For testing we should leave it as it is, but afterwards we should change to more commonly used systems as we're going to pack everything into an npm package (maybe with prebuilt binaries ?)

**Comment by joshgarde** [comment]
I've found some good demo code under `src/ui/aura/demo/demo_main.cc`

**Comment by joshgarde** [comment]
Down the line I was also thinking of replacing the full Chromium implementation in Ash with the embedded version since it's a lot easier to build with a precompiled library than to keep building the entire Chromium source all the time.

**Comment by joshgarde** [comment]
I've found some good demo code under `src/ui/aura/demo/demo_main.cc`

Down the line I was also thinking of replacing the full Chromium implementation in Ash with the embedded version since it's a lot easier to build with a precompiled library than to keep building the entire Chromium source all the time.

**Comment by joshgarde** [comment]
More useful demo code:

```
src/ui/ozone/demo/ozone_demo.cc
src/ui/views/examples/examples_main.cc
```

**Comment by Coretool** [comment]
Thanks for the code, I'll look into it!

I think we should build the content shell with ozone first and then build everything ontop of that. That's probably like what you suggested :).

**Comment by joshgarde** [comment]
Yep, that's what I had in mind

**Comment by joshgarde** [comment]
Quick update: Instead of trying to modify the source directly and resolving dependencies independent of Google scripts, I'm going to be modeling the project after NodeOS/qemu. So instead of having the source in repo, it'll be downloaded along with all the other tools and dependencies when a build is run. We'll be building aurora on top of the existing source.

**Comment by Coretool** [comment]
Yeah, sure, that's what I thought we would do anyway ^^

**Comment by Coretool** [comment]
However, we should make it optional because it takes quiet some time to build it...

**Comment by joshgarde** [comment]
Yeah - primarily we're going to use prebuilds

**Comment by Coretool** [comment]
So we will have to create a script that builds prebuilds for every platform ...

**Comment by joshgarde** [comment]
Only Linux and since it's the early development stages, we'd only need to target 64-bit.

**Comment by Coretool** [comment]
Yeah but for the future since NodeOS will target 32-bit, arm, etc ^^

**Comment by joshgarde** [comment]
Got the aurora demo compiled, but I can't get it to run on my Ubuntu box - Possible issue: https://groups.google.com/a/chromium.org/d/msg/ozone-dev/PrqBxBmSG3s/Qyv_GZKtAwAJ

**Comment by Coretool** [comment]
Did you terminate X ? 

I think the first two we will be able to patch ... the third one will be a bit more challenging but should be possible

**Comment by joshgarde** [comment]
I didn't even start X - Got to my display manager's login ui and immediately went into the CLI - Got a similar error to the OP of the topic.

I did get it "running" with X running, but with no display on screen.

**Comment by joshgarde** [comment]
The first two I think we already have. The last one might require some patching...

**Comment by Coretool** [comment]
I'll search the driver / export it from ChromeOS

**Comment by joshgarde** [comment]
https://chromium.googlesource.com/chromiumos/third_party/libdrm/
^ This is probably the one we're after

https://chromium.googlesource.com/chromiumos/third_party/mesa/

**Comment by joshgarde** [comment]
I've had success compiling libdrm on my Ubuntu VPS without the need to checkout the entire ChromiumOS repo (just needed to install libpciaccess-dev). I'm going to try installing it on my local box tomorrow.

**Comment by Coretool** [comment]
So do we still need point 3 ?

**Comment by Coretool** [comment]
I'm cloning the whole chromium src now and will grap what I need to start hacking on it. Do you mind, if we list what we have to do and afterwards split it into several issues so we can track progress on everything ?!

**Comment by joshgarde** [comment]
Sure - libdrm should resolve point 3, but I still need to test

**Comment by Coretool** [comment]
So @joshgarde did you get the content shell ?

**Comment by joshgarde** [comment]
Either I'm not installing the library correctly or something else is needed -

I'm going to test a few more things out, but some other repos to probably investigate:
https://chromium.googlesource.com/chromiumos/third_party/drm/ (<-- Seems outdated)
https://chromium.googlesource.com/chromiumos/drm_hwcomposer/

**Comment by Coretool** [comment]
Why do we not just copy the driver out of a fully functional chromeos image (i.e. use a vm and then extract the driver ?)

**Comment by joshgarde** [comment]
We can, but I would also like to be able to build from source if the binary isn't available (this is an open source project after all).

**Comment by Coretool** [comment]
Yeah... maybe we want to contact a chromium dev ? (Using the mailing list )

**Comment by joshgarde** [comment]
Sure - it'd speed up development a bit

**Comment by Coretool** [comment]
So I'll go and make some friends... or try to ?! ^^

**Comment by joshgarde** [comment]
Go for it - I'll join in too

**Comment by Coretool** [comment]
Have you seen / tried out the 'tutorial' on how to deploy an ozone content_shell ? It looks pretty good ... 

What did you compile exactly ?

**Comment by joshgarde** [comment]
https://www.chromium.org/developers/design-documents/ozone

Truthfully, I didn't see that part of the documentation. 😪 I was compiling aura_demo.

**Comment by Coretool** [comment]
Yeah, I thought that ^^ I think we should try this one out (I'll do that soon)

**Comment by joshgarde** [comment]
Compiling right now on my VPS - Will report back with results

**Comment by Coretool** [comment]
@joshgarde Any results yet ?

**Comment by joshgarde** [comment]
First compile failed because it was compiled with X dependencies. I'm doing a second compile now, but it's taking more than an hour. ~~The final compiled binary is about 1.4GB with about 700MB of libraries and other files.~~

EDIT: Final binary & support files without X deps is only 1GB.

**Comment by joshgarde** [comment]
Still the same error I'm afraid -

```
libEGL warning: DRI3: xcb_connect failed
libEGL warning: DRI2: xcb_connect failed
eglInitalize Default failed with error EGL_NOT_INITALIZED
```

**Comment by joshgarde** [comment]
Ubuntu 16.04 with i965 integrated graphics

**Comment by joshgarde** [comment]
Going to attempt install on a clean system

**Comment by joshgarde** [comment]
[content_shell.txt](https://github.com/joshgarde/aurora/files/367707/content_shell.txt)
^ Full debug log

I've forced EGL to render to drm which makes `libEGL warning: DRI2: xcb_connect failed` disappear, but that's about it.

Going to attempt a second run on my older (and fresher) Debian box.

**Comment by Coretool** [comment]
Xcb... Always those Xs :/ Nothing new from the chromium guys either !

**Comment by Coretool** [comment]
@joshgarde Could you push your Code ? I might have the solution !!

**Comment by piranna** [comment]
Any update on this guys? Seems a nice work! :-)

**Comment by joshgarde** [comment]
Chromium's codebase is a mammoth. It takes a while just to compile. Last summer's efforts were halted when I managed to screw up the display server on my laptop and wasn't able to fix it because I was going on a vacation the next day.

I still want to use Chromium OS' display system because (once it's done compiling), it's pretty lightweight. I might pick this back up later.

**Comment by piranna** [comment]
Just to clarify: Aurora is the graphic system / web rendered (like X11 or noGUI) and ash is the window manager, isn't it? Maybe both should be splitted in two independent projects... Have you tried to make ash to run on a regular web browser to a remote NodeOS machine?

OffTopic: my ex-girlfriend's name was Aurora... that's ironic :-P

**Comment by joshgarde** [comment]
Aura is the low-level display system for Chromium OS. Chromium OS uses Ash on top of Aura for Chromium UI stuff like window management, desktop environment, etc.

Aurora is the name I chose for my NodeOS gui project although I am considering changing it to noDesktop

**Comment by Coretool** [comment]
...and chromium itself is used to render the UI afaik

**Comment by joshgarde** [comment]
Just picked up a VPS and started compiling Chromium (With Aura, Ash, and Ozone-GBM support). It's been running for almost 12 hours now, but I have a feeling that's because I got a 1vCPU, 4GB RAM VPS.

**Comment by joshgarde** [comment]
Look's like the Chromium team is helping us out here though -
https://www.chromium.org/developers/mus-ash

> a project to separate the window management and shell functionality of ash from the chrome browser process

**Comment by Coretool** [comment]
But we will need ash and toolkit_views first since mus+ash depends on them

**Comment by piranna** [comment]
I start to get a bit lost here... Does Chromium run directly on top of EGL and have a HTML5+JS window manager and/or desktop (Ash)? Or on the other hand, is Ash a regular window manager and/or desktop like metacity/gnome that can only be able render Chromium browser windows, and Aura is like Wayland / X11? If that's the point, then we'll have a problem for portability...

My idea for the GUI is to use a web renderer (Chromium, Gecko, regular web browser on a desktop computer) as graphic system (like X11), and run on it a HTML5+JS window manager and desktop, and the application windows being iframes. I think this path will be fairly easier to achieve (we can start working on it right now, and probably we could port current ones like `OS.js` or `ember-shell`) and would have portability warranted.

**Comment by joshgarde** [comment]
Chromium has it's own renderer (Aura) which it can use to talk directly to video devices though the kernel's DRM api. Ash is the window manager (Which uses chromium to render it's content) that runs on top of Aura.

My main concern with going the iFrame route is that there's no separation between the process that runs the desktop and the process that runs the applications. They're the same process. A bad script executed in an application can crash the entire desktop environment.

As a side note - I just got my i7 box loaded up with Ubuntu and the chromium src tree. 😄  Builds have never felt so fast.

**Comment by piranna** [comment]
> My main concern with going the iFrame route is that there's no separation between the process that runs the desktop and the process that runs the applications. They're the same process. A bad script executed in an application can crash the entire desktop environment.

I don't know if it would crash just the bad iframe or the full desktop... I know that there was work on making each iframe to run optionally in their own process. Probably this is what Gecko did with the `<browser>` tag (an extension of iframe) on FirefoxOS.


> As a side note - I just got my i7 box loaded up with Ubuntu and the chromium src tree.  Builds have never felt so fast.

Nice! :-D

**Comment by Coretool** [comment]
We could *borrow* the webview implementation from electron which provides a "more than usual" grade of isolation ...

**Comment by piranna** [comment]
Don't know how much complicate would be to create something new than using things off-the-shelf...
